### ISSUE TITLE
How the sample_weight parameter is used inside the fit function of lgb.LGBMRegressor

### ISSUE THREAD

**Issue by ZLake** [Microsoft/LightGBM#905]
## Environment info
Operating System: Mac OS EI Capitan 10.11.6 (15G31)
CPU:1.6 GHz Intel Core i5
C++/Python/R version: python 3.6.1

## Thanks in advance!!!

**Comment by Laurae2** [comment]
> when I set all sample weight to 2, it gives different result as the previous case with all sample weight equals 1

This is the expected behavior, as all gradients and hessians sums are mulitplied by 2. It incurs a different model as regularization / pruning is not performed the same, if not scaled appropriately.

**Comment by ZLake** [comment]
@Laurae2 thanks for ur reply, it makes total sense to me, awesome!

Is there any recommended way of setting the weight, like the largest weight is 1, the rest sample weight decreases based on their importance compared with the most important samples.

BTW, if this is the case, what is the trained model when the sample weight is 0? Because I also want to disable the effect of some sample by setting its weight to 0, but seems that even its weight is 0, it contributes to the model?

**Comment by Laurae2** [comment]
> Is there any recommended way of setting the weight, like the largest weight is 1, the rest sample weight decreases based on their importance compared with the most important samples.

It depends on your dataset and your loss function.

> BTW, if this is the case, what is the trained model when the sample weight is 0? Because I also want to disable the effect of some sample by setting its weight to 0, but seems that even its weight is 0, it contributes to the model?

Can you provide a reproducible result?

**Comment by Laurae2** [comment]
Note that for the "weight = 0" issue, LightGBM may perform differently the binning (if you remove the observation).

This can be observed with the following:

```r
# raw data
[LightGBM] [Info] Total Bins 68
[LightGBM] [Info] Number of data: 100, number of used features: 52
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[1]:	test's l2:0.376768 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=3 and max_depth=2
[2]:	test's l2:0.380581 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[3]:	test's l2:0.384609 

# raw data with one zero weight
[LightGBM] [Info] Total Bins 68
[LightGBM] [Info] Number of data: 100, number of used features: 52
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[1]:	test's l2:0.376768 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[2]:	test's l2:0.380581 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[3]:	test's l2:0.384609 

# raw data with zero weight removed
[LightGBM] [Info] Total Bins 68
[LightGBM] [Info] Number of data: 99, number of used features: 52
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[1]:	test's l2:0.382543 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[2]:	test's l2:0.385894 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[3]:	test's l2:0.389485 

```

Reproducible code:

```r
lgb.unloader(wipe = TRUE)
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
train$label <- train$label[1:100]
train$data <- train$data[1:100, ]
dtrain <- lgb.Dataset(train$data, label = train$label, weight = rep(1, 100))
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = "l2", min_data_in_leaf = 1)
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   3,
                   valids,
                   min_data = 1,
                   learning_rate = 0.1,
                   early_stopping_rounds = 10)

lgb.unloader(wipe = TRUE)
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
train$label <- train$label[1:100]
train$data <- train$data[1:100, ]
dtrain <- lgb.Dataset(train$data, label = train$label, weight = c(rep(1, 99), 0))
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = "l2", min_data_in_leaf = 1)
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   3,
                   valids,
                   min_data = 1,
                   learning_rate = 0.1,
                   early_stopping_rounds = 10)

lgb.unloader(wipe = TRUE)
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
train$label <- train$label[2:100]
train$data <- train$data[2:100, ]
dtrain <- lgb.Dataset(train$data, label = train$label, weight = rep(1, 99))
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = "l2", min_data_in_leaf = 1)
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   3,
                   valids,
                   min_data = 1,
                   learning_rate = 0.1,
                   early_stopping_rounds = 10)
```

**Comment by ZLake** [comment]
@Laurae2 Thank you for ur reply, it helps a lot! Really great answer!
